### PR TITLE
 src/mds: increment directory inode's change attr by one

### PR DESCRIPTION
**Pull request by ajarr**
... whenever the mtime or ctime of the directory inode is modified.
    
    In CephFS subvolume clones exported using NFS-Ganesha, newly created
    files using `touch` were not being listed. It was identified that the
    create request sent to the Ceph MDS via NFS-Ganesha's libcephfs client
    modified the mtime and ctime of the parent directory, but did not modify
    the change_attr of the parent directory. Since the NFS client
    didn't see a modification of the change attribute in the reply, it
    didn't invalidate its readdir cache. The subsequent directory `ls` was
    satisfied from the NFS client's stale readdir cache.
    
    Whenever parent directory inode's  mtime was modified in
    MDCache::predirty_journal_parents(), the parent inode's change_attr
    was set to its dirstat->change_attr. The parent inode's
    dirstat->change_attr doesn't track changes to parent's *ctime only*
    changes such as setattr, setvxattr, etc. on the parent
    directory. See commit 0d441dcd6af553d11d6be6df56d577c5659904a0 for more
    details. This caused the directory inode's change_attr to not be updated
    when an operation to change only its ctime was followed by an operation
    to change its mtime and ctime.
    
    Fix this by making changes to MDCache::predirty_journal_parents() and
    CInode::finish_scatter_gather_update() to increment the directory
    inode's change_attr by one instead of setting it to its
    dirstat->change_attr.
    
    Fixes: https://tracker.ceph.com/issues/57210
    Signed-off-by: Ramana Raja <rraja@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [x] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by gregsfortytwo** [comment]
Nice find tracking down the cause of this bug!

The biggest concern here is just making sure we don't have any outstanding places where we set the change_attr to the value of another dirstat or something, because that could rewind us back to having this bug again. Sounds like that's been verified @ajarr? :)

**Comment by ajarr** [comment]
> Nice find tracking down the cause of this bug!
> 
> The biggest concern here is just making sure we don't have any outstanding places where we set the change_attr to the value of another dirstat or something, because that could rewind us back to having this bug again. Sounds like that's been verified @ajarr? :)

I tracked down the two instances where directory inode's change_attr was set to its dirstat's change_attr, and made the required changes in this PR.  Not much has changed around change_attr, since the PR https://github.com/ceph/ceph/pull/10922/ that introduced it.

**Comment by ajarr** [comment]
@jtlayton I looked at your PR that introduced change_attr https://github.com/ceph/ceph/pull/10922. I don't see much code changes around _change_attr_ tracking since then in Ceph. In this PR, I make changes in the MDS code where directory inode's change_attr were set to its _dirstat's change_attr_ (by https://github.com/ceph/ceph/pull/10922/commits/0d441dcd6af553d11d6be6df56d577c5659904a0), and instead increment the directory inode's change_attr by one as explained in the commit message.

**Comment by ajarr** [comment]
@vshankar, these are operations that happen on a FS subvolume clone export directory (/volumes/<group name>/<clone name>/UUID/) when you issue `fs subvolume snaphsot clone` command.

mkdir 
setxattr 'ceph.dir.subvolume' on directory
setattr mode=0755 on directory
files copied over from snapshot created within export directory
setattr mtime, atime on directory

**Review by vshankar — APPROVED** [review]
Great catch!

**Review by kotreshhr — APPROVED** [review]

**Comment by gouthampacha** [comment]
This is great! Thank you @ajarr - looking forward to backports when you have them; and i'll be happy to test with OpenStack in the picture!

**Comment by vshankar** [comment]
@ajarr you mentioned about an alternate fix. Do you plan to work on it and discuss it here? (A separate PR for that would work too).

**Comment by ajarr** [comment]
@vshankar I plan to work on it in a separate PR. This PR can be merged if the tests passed.

File: src/test/libcephfs/test.cc [hunk 1]
```diff
@@ -1889,17 +1889,51 @@ TEST(LibCephFS, DirChangeAttr) {
...
+  old_change_attr = stx.stx_version;
+
+  /* chmod dir, and check whether dir's change_attr changed */
+  ASSERT_EQ(ceph_chmod(cmount, dirname, 0777), 0);
+  ASSERT_EQ(ceph_statx(cmount, dirname, &stx, CEPH_STATX_VERSION, 0), 0);
+  ASSERT_TRUE(stx.stx_mask & CEPH_STATX_VERSION);
+  ASSERT_NE(stx.stx_version, old_change_attr);
+  old_change_attr = stx.stx_version;
 
+  /* rename a file within dir, and check whether dir's change_attr changed */
```
Comment (batrick) [inline src/test/libcephfs/test.cc:1920]: Can you break these logically distinct tests into separate unit tests?

**Review by batrick — CHANGES_REQUESTED** [review]
Great work!

File: src/test/libcephfs/test.cc (hunk 1 above)
Comment (ajarr) [reply src/test/libcephfs/test.cc:1952]: Done.

**Comment by ajarr** [comment]
In the latest patchset, only made changes to the test to break it down into smaller ones.

**Comment by vshankar** [comment]
jenkins test api

**Review by batrick — APPROVED** [review]

File: src/test/libcephfs/test.cc [hunk 2]
```diff
@@ -1870,40 +1870,141 @@ TEST(LibCephFS, ChangeAttr) {
...
-  ASSERT_EQ(ceph_statx(cmount, dirname, &stx, CEPH_STATX_VERSION, 0), 0);
+  ASSERT_EQ(ceph_statx(cmount, dirpath, &stx, CEPH_STATX_VERSION, 0), 0);
   ASSERT_TRUE(stx.stx_mask & CEPH_STATX_VERSION);
 
   uint64_t old_change_attr = stx.stx_version;
 
-  int fd = ceph_open(cmount, filename, O_RDWR|O_CREAT|O_EXCL, 0666);
+  /* Alternate operation that changes directory's ctime (setxattr)
+   * with one thats change directory's mtime and ctime (create).
+   * Check that directory's change_attr is updated everytime ctime changes.
```
Comment (batrick) [inline src/test/libcephfs/test.cc:1894]:
```suggestion
  /* Important: Follow an operation that changes the directory's ctime (setxattr)
   * with one that changes the directory's mtime and ctime (create).
   * Check that directory's change_attr is updated everytime ctime changes.
```

Comment (ajarr) [reply src/test/libcephfs/test.cc]: Made this change. Thanks!

**Comment by ajarr** [comment]
In the latest patchset made the following changes to the tests:
- improved readability of the code comments per Patrick's suggestion
- checked that change_attr increases everytime ctime is updated. So used  ASSERT_GT instead of ASSERT_NE

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#14-Oct-2022

**Comment by vshankar** [comment]
jenkins test api

**Pull request merged**